# Module 1. Samples

## Learning goals

- You are able to use Jupyter Notebooks (locally or in Google Colab) and configure your environment to run the code examples in this course.
- You understand the basic terminology used in data science (indicated with 💡).
- You are able to load a dataset in Python and explore its structure and characteristics.

## Getting Started with Jupyter Notebooks

Jupyter Notebooks (or iPython Notebooks) are files with a mixture of Markdown and Python code. It's a very convenient format so we'll be using this quite a lot during this course!

**Read the instructions below carefully, as they do not necessarily apply to all environments!**

### Google Colab

To get started, it is not necessary to install software on your laptop. [Google Colab](https://colab.research.google.com/) is a free online iPython development environment, but it does require a Google account in order to use it. You can create a local clone of the Github repo with the lab assignments and upload the Notebook files you want to work with to Colab.

Note that the Google Colab environment has older versions installed of the Python libraries that we will be using throughout this course. Usually, that's not a problem, but sometimes you will run into errors, e.g. because some parameter in our Python code didn't exist yet in that version. The solution is to add a new code block at the beginning of your Notebook file with content:

In [ ]:
%pip install --upgrade matplotlib
%pip install --upgrade scipy
%pip install --upgrade statsmodels

You only have to do this in Notebooks where you encounter errors, and only for the libraries that cause these errors. The installation may take a while and you need to restart the Python runtime afterwards by clicking the button that appears.

### Local environment

If you prefer to have a local development environment, you need to install Python and Visual Studio Code with the necessary extensions for working with Python (e.g. Pylance, Jupyter, Jupyter Keymap, etc.) The first time you open a Notebook file, VSCode will offer you to install these extensions.

### Windows
To install Python and VSCode on Windows, we recommend using `winget`. First, check the latest version of Python on the [Python website](https://www.python.org/downloads/windows/). For example, at the time of writing, the latest version was 3.14.2. We'll need to specify the major and minor version (e.g. 3.14) when installing. Then, open a PowerShell or command prompt as Administrator and run:

```console
> winget install Microsoft.VisualStudioCode
> winget install Python.Python.3.14
```

Remark that `winget upgrade --all` will install maintenance releases of Python, but it will not upgrade to a new major or minor version.

### Linux

Some Linux distributions no longer allow you to use `pip` to install Python packages globally. In that case, we recommend to work with a Python Virtual Environment. VS Code will help you to create one when you select the Python version to use for executing code cells.

### MacOS

The install Python on MacOS, it's best to use [Homebrew](https://brew.sh/). After installing issue the following commands.

```console
> brew install python
> export PATH="/opt/homebrew/opt/python/libexec/bin:$PATH"
```
You probably want to add the `export PATH` line to your `.zshrc` (or whichever shell you use)

You can check everything is correctly set through

```console
> which python
/opt/homebrew/opt/python/libexec/bin/python
```

It's always a good idea to set up a venv

```console
dsai-labs> python3 -m venv .venv
source .venv/bin/activate
```

You should now be able to use this environment as your kernel environment in VSCode.

## Libraries

Most Python scripts (or Jupyter Notebooks) for analyzing data use the same program libraries. The most important ones are:

- `numpy` - multidimensional arrays, linear algebra, etc.
- `scipy` - mathematics, science, engineering etc.
- `pandas` - data-analysis and -manipulation
- `matplotlib`, `seaborn` - data visualisation
- `scikit-learn` - regression and machine learning

You can install these using `pip`:

```console
> pip install matplotlib numpy pandas scipy seaborn scikit-learn statsmodels
```

Or, alteratively, by executing the following code block:

In [ ]:
%pip install matplotlib
%pip install numpy
%pip install pandas
%pip install scipy
%pip install seaborn
%pip install scikit-learn
%pip install statsmodels


Since you usually need the same packages every time, it is best to put them at the top of every script or Notebook file you write. By convention, package names are often abbreviated in a consistent manner, e.g. `np` for `numpy`, `sns` for `seaborn`, etc.

In [2]:
# Importing the necessary packages
import numpy as np                                  # "Scientific computing"
import scipy.stats as stats                         # Statistical tests

import pandas as pd                                 # Data Frame
from pandas.api.types import CategoricalDtype

import matplotlib.pyplot as plt                     # Basic visualisation
from statsmodels.graphics.mosaicplot import mosaic  # Mosaic diagram
import seaborn as sns                               # Advanced data visualisation

**Tip:** you can define a code snippet in VSCode for this block of code, so you don't have to type it every time. See e.g. [this guide](https://code.visualstudio.com/docs/editor/userdefinedsnippets) on how to create code snippets in VSCode.

## Basic Concepts in Data Science

### Variables and Measurement Levels

In statistics and data science, a 💡**variable** is any attribute of an object of study. Variables can have 💡**values** that differ between objects. Usually, variables are classified into different types, or 💡**measurement levels**, depending on the kind of values they can take and the mathematical operations that can be performed on them. The most common types are:

- 💡**Qualitative** (or categorical) variables: variables that can take on a limited number of distinct categories or labels. These are not necessarily numerical.

    - 💡**Nominal** variables: categorical variables *without a specific order* or ranking among the categories (e.g., colors, types of animals).
    - 💡**Ordinal** variables: categorical variables *with a specific order* or ranking among the categories (e.g., education level, customer satisfaction ratings).

- 💡**Quantitative** (or numerical) variables: variables that can take on numerical values, which can be discrete (countable values) or continuous (any value within a range). These are usually the result of measurements. Often, there are many different values of quantitative variables in a dataset, or they may even be unique.

    - 💡**Interval** variables: numerical variables where the intervals between values are meaningful, but there is no true zero point (e.g., temperature in Celsius or Fahrenheit).
    - 💡**Ratio** variables: numerical variables with a true zero point, allowing for meaningful ratios between values (e.g., height, weight, age).

It is important to correctly identify the type of variable you are working with, as this influences the choice of statistical methods and visualizations used for analysis. This is something that a computer is not very good at, so you often need to make these distinctions yourself when exploring and analyzing a dataset.

### Populations, samples, observations

In data science and statistics, a 💡**population** refers to the entire set of individuals or items that you are interested in studying. A 💡**sample** is a subset of the population that is selected for analysis. A single individual or item within a population or sample is called an 💡**observation**.

The goal of sampling is to gather information about the population without having to study every individual or item within it. Under certain conditions (which will be explored later in this course), you can use the information obtained from a sample to make inferences about the entire population. This is often necessary because studying the entire population can be impractical, time-consuming, or expensive.

When selecting a sample from a population, it is important to ensure that the sample has specific characteristics. Specifically, a sample should be:

- 💡**Randomly selected**, i.e. every individual or item in the population should have an *equal chance of being included* in the sample in order to reduce bias, but randomness alone does not guarantee that a sample is perfectly representative - especially for small sample sizes..
- **Large enough** to provide reliable estimates of population parameters. We will explore this more in chapter 3.
- 💡**Representative** of the population, meaning that it should accurately reflect the characteristics of the population as a whole.

Ideally, a researcher would do the following to draw a sample from a population:

- First and foremost, the population must be well defined
- Then you can compile a 💡**sampling frame**, i.e. a list of all individuals in the population
- Then, the researcher randomly selects a number of individuals from this list.

It is often difficult or impossible to set up a sampling frame. In such cases, the researcher chooses a different sampling method and tries to approximate the properties listed above as closely as possible within the available time and budget.

### Sampling errors


Even under the best of circumstances, the result within a sample is likely to deviate from the "true" value in the population, although we hope we can approach it sufficiently. So by taking a sample we inevitably introduce errors. A useful way to categorize errors is given below:

- 💡**Sampling errors** are caused by the way we take a sample. We distinguish between 💡*random* or 💡*systematic* sampling errors:

    - **Random sampling errors** are only due to chance and cannot be avoided. Every time you take a random sample, you will get a different result within that sample.

    - **Systematic sampling errors** are worse, because they can cause the results to be biased and therefore no longer be *representative* of a population. Examples:

        - Online survey: people without internet are excluded

        - Street survey: only people who are currently walking there

        - Voluntary survey: only interested parties participate

- 💡**Non-sampling errors** are errors that occur after the sample has been taken, e.g. in the measurement and analysis. In this case, we speak of non-sampling errors and again we distinguish between *random* and *systematic*:

    - **Random non-sampling errors** For example, a respondent accidentally selects or fills out a wrong answer in a survey.

    - **Systematic non-sampling errors** For example:

        - a measuring device that is not calibrated correctly and systematically gives a slightly higher value than the true value. 

        - respondents intentionally underestimate or overestimate reality (e.g. when asked about how much you smoke or drink alcohol on a daily basis)

### Representative samples

A sample is said to be 💡**representative** of a population when the characteristics of the sample closely match those of the population. This means that the sample accurately reflects the diversity and distribution of key variables within the population, allowing for valid inferences to be made about the entire population based on the sample data.

In a 💡**stratified** sample, the population is divided into subgroups (or strata) based on certain characteristics, and samples are drawn from each stratum in proportion to their representation in the population. This helps ensure that the sample is representative of the population as a whole, especially when certain subgroups may be underrepresented in a simple random sample.

For example, if a population of 5810 individuals consists of the following age groups and genders:

|           | $\le 18$ | $]18,25]$ | $]25, 40]$ | $> 40$ | Total |
| --------: | -------: | --------: | ---------: | -----: | ----: |
|     Woman |      500 |      1500 |       1000 |    250 |  3250 |
|       Man |      400 |      1200 |        800 |    160 |  2560 |
| **Total** |      900 |      2700 |       1800 |    410 |  5810 |

A stratified sample of 10% would then consist of 581 individuals, with the following distribution:

|           | $\le 18$ | $]18,25]$ | $]25, 40]$ | $> 40$ | Total |
| --------: | -------: | --------: | ---------: | -----: | ----: |
|     Woman |       50 |       150 |        100 |     25 |   325 |
|       Man |       40 |       120 |         80 |     16 |   256 |
| **Total** |       90 |       270 |        180 |     41 |   581 |

In chapter 4, we will discuss a method that allows us to verify whether a sample is representative of a population.

### Univariate vs multivariate analysis

In chapter 2, on 💡**univariate analysis**, we will focus on analyzing and summarizing data for a single variable at a time. This involves techniques such as calculating measures of central tendency (*mean*, *median*, *mode*) and measures of dispersion (*range*, *variance*, *standard deviation*), as well as creating visualizations like histograms and box plots to understand the distribution of the variable.

In chapters 4 to 6, we will explore various methods to investigate whether two variables have some kind of relationship or association with each other, under the umbrella term of 💡**bivariate analysis**.

The appopriate analysis or visualization technique always depends on the measurement levels of the variables involved. For example, we might use a scatter plot to visualize the relationship between two quantitative variables, or a contingency table to examine the association between two categorical variables.

We say that two variables have an 💡**association** or a 💡**relationship** when the value of one variable provides some information about the value of the other variable. For example, there may be an association between age and income, where older individuals tend to have higher incomes. The variable that provides information about the other variable is called the 💡**independent variable**, while the variable being predicted or explained is called the 💡**dependent variable**.

It is essential to understand that when you find an association between two variables, it does not necessarily imply a causal relationship. A phrase often used in this context is **correlation is not causation**. In other words, just because two variables are associated does not mean that one variable causes the other to change. There may be other factors at play, or the association could be entirely coincidental (*spurious correlation*). See some examples [here](https://plotlygraphs.medium.com/spurious-correlations-56752fcffb69) and [here](https://www.tylervigen.com/spurious-correlations).

## Opening a Dataset, General Information

Throughout this course, we will load 💡**datasets** from various sources to practice our data analysis skills. A dataset usually corresponds to a *sample* taken from a *population* being studied, and is typically organized in a tabular format where rows represent individual *observations* and columns represent different *variables*.

To read a dataset into Python, you can use the `pandas` library, which provides powerful tools for data manipulation and analysis. Pandas has several functions to read data from various file formats, such as CSV (`read_csv()`), Excel (`read_excel()`), JSON (`read_json()`), and more.

You can specify a path to a local file or even a URL to download the dataset directly from the web:

In [3]:
# Importing the Titanic dataset. (Rajagopalan, 2021, p. 106)
titanic = pd.read_csv('https://raw.githubusercontent.com/DataRepo2019/Data-files/master/titanic.csv')
# Show the first few records of the Data Frame
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

If you want to open a dataset that you stored on Google Drive (where Google Colab stores all your Notebooks), you can use the following code instead:

In [ ]:
# Ensure your Google Drive is accessible from within the notebook
from google.colab import drive
drive.mount('/content/drive')

# Load the dataset from your Google Drive (here, we stored it in a subfolder
# called "data")
titanic = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/data/titanic.csv')
titanic.head()

Let's take a look at the properties of the dataset we just loaded:

In [4]:
# How many  rows does the DataFrame have?
print(f"Number of rows: {len(titanic)}")
# How many columns?
print(f"Number of columns: {len(titanic.columns)}")
# How many rows and columns, i.e. the shape
print(f"The shape of the Data Frame is: {titanic.shape}")
# General information about the DataFrame
print("*"*50)
titanic.info()

# Give the data type of each column.
print("*"*50)
print(titanic.dtypes)

# How many columns of each data type are there?
#   Watch it! The book says to use get_dtype_counts(), but this method no longer exists
print("*"*50)
print(titanic.dtypes.value_counts())

Number of rows: 891
Number of columns: 12
The shape of the Data Frame is: (891, 12)
**************************************************
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
**************************************************
PassengerId      int64
Survived         int64
Pclass      

### Indices

The column "PassengerId" is not an actual variabele, but contains a number to identify each observation. You can mark this column as an index:

In [ ]:
titanic.set_index(['PassengerId'])

### Qualitative variables

Earlier, we stated that the computer is not always able to correctly identify the measurement levels of variables. In this dataset, some of the variables, such as `Survived` and `Pclass`, are incorrectly considered to be quantitative. You can correct this by explicitly converting them to a **qualitative** (categorical) variable:

In [ ]:
# Describe the variable Survived -> is treated as quantitative
print(titanic.Survived.describe())
# Convert to a categorical variable
titanic.Survived = titanic.Survived.astype('category')
# Call desribe() again -> now it is treated as qualitative
print(titanic.Survived.describe())

You can also mark variables as **ordinal**, that is, with an ordering. We will do this as an example with the variable "Embarked" and order the ports by the order in which the Titanic stopped there. The Titanic departed at Southhampton, and then picked up passengers first at Cherbourg and then at Queenstown.

For cases like this, define your own datatype specifying the order:

In [ ]:
print(titanic.Embarked.unique())

embarked_type = CategoricalDtype(categories=['S', 'C', 'Q'], ordered=True)
titanic.Embarked = titanic.Embarked.astype(embarked_type)
titanic.Embarked.describe()

This order will then always be respected, e.g. in graphs

In [ ]:
sns.countplot(data=titanic, x='Embarked');

### Selecting Data

In [ ]:
# Select all observations for a single variable (i.e. a DataFrame column)
titanic.Age
# This also works (and is preferable as it will also work when the column name has a space in it):
# titanic['Age']
# This also works, but isn't very nice
# titanic.loc[:, 'Age']

In [ ]:
# Select adjacent columns
titanic.iloc[:, 2:4]

You can also select multiple columns based on their names.
This is often clearer than selecting based on position and the columns must 
not be adjacent.

In [ ]:
titanic[['Name', 'Age', 'Cabin']] # Note: two sets of square brackets!

In [ ]:
# Observation with row number 5 (counting from zero)
print(titanic.iloc[5])

# The first 4 observations
titanic.iloc[0:4]

In [ ]:
# Select observations where the value of Age is less than 18
titanic[titanic.Age < 18]  

# The same, but only keep the column 'Embarked'
titanic[titanic.Age < 18].Embarked

# The same, but keep columns 'Age' and 'Embarked'
titanic[titanic['Age'] < 18][['Age', 'Embarked']]

In [ ]:
# Select all boys younger than 10
titanic.query("(Sex=='male') and (Age < 18)")

### Dropping Data and Working with Missing Data

Pandas tries to make working with missing data as easy as possible. E.g., all of the descriptive statistics on pandas objects exclude missing data by default.

Let's start by reading the titanic data again.

In [ ]:
titanic = pd.read_csv('https://raw.githubusercontent.com/DataRepo2019/Data-files/master/titanic.csv')
titanic.head()

The `PassengerId` column is identical to the index hence it doesn't provide any additional useful information. Let's drop this column (rather than setting it as the index as before). This can be done using `drop`.

In [ ]:
titanic.drop("PassengerId", axis="columns")

Notice that we didn't actually change the `titanic` `DataFrame`.  We can either assign the result of `drop` to a (new) `DataFrame`
or we can use `inplace=True`.

In [ ]:
titanic.head(3) # PassengerId is still here

In [ ]:
titanic = titanic.drop("PassengerId", axis="columns")
titanic.head(3) # PassengerId is now gone

From the output of the `info` method we can infer that certain columns contain many missing values, e.g. `Cabin` only contains 204 non-missing values. 
For the purposes of illustration, let's drop any row that has a missing observation by using 
the default behaviour of the `dropna` method.

In [ ]:
titanic.info()

In [ ]:
cleaned = titanic.dropna() # Drop any row that has at least one missing value
print(cleaned.info())
cleaned.head(5)

This would only keep 183 values, which isn't a lot. We could also choose to drop only the rows that consist of nothing but missing values (there are no such rows in the `titanic` `DataFrame`), but here you can see how this would be done.

In [ ]:
cleaned = titanic.dropna(how="all")
cleaned.info()

Instead of relying on `info` to compute the number of non-missing values we can also use the `isnull` and/or `notnull` methods.
`notnull` computes a boolean `DataFrame` where an entry is `True` if and only if that entry is considered to be non-missing in the original `DataFrame`. We can then use `sum` to compute the sum of each column. This gives the number of non-missing values per column.  

In [ ]:
not_null_df = titanic.notnull()
print(not_null_df.tail(3))
print("Number of non null values in each column:")
print(not_null_df.sum())

An even easier alternative is to simply use `count` on the `DataFrame`.

In [ ]:
print(titanic.count())

The `Cabin` column has too many missing values, so probably isn't useful. For the `Age` column, we can *impute* the missing values, e.g. with the average age of all passengers. (We will see later what this means exactly). Let's do this using `fillna`.

In [ ]:
# First compute the average age
avg_age = titanic['Age'].mean()
print(f"(Rounded) Average age of passengers: {round(avg_age)}")
titanic = titanic.fillna(value={'Age' : avg_age})
print("We should now confirm that the 'Age' column no longer has missing values")
titanic.info()

### Creating New Columns

We use a dataset containing $NO_2$ measurements at the stations of Paris, Antwerp and London.

Let's read the data and check whether it worked.

In [ ]:
URL = "https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/air_quality_no2.csv"
air_quality = pd.read_csv(URL)
print(air_quality.info())
air_quality.head()

We see that 
- the dates are stored as strings (object dtype)
- `datetime` is added as a column and we have a `RangeIndex`.

We would like to
- parse the dates, so that they are available as `datetime` objects
- index the `DataFrame` rows by this date.

This can be achieved by specifying the index column and by indicating that dates must be parsed on `read_csv`. Let's try again.

In [ ]:
air_quality = pd.read_csv(URL, index_col=0, parse_dates=True)
print(air_quality.info())
air_quality.head()

We can clearly see the difference!

### Creating a New Column Derived from (an) Existing Column(s)

Assume we would like to express the concentration of the station in London in mg/m^3

*(If we assume temperature of 25 degrees Celsius and pressure of 1013 hPa, the conversion factor is 1.882)*

In [ ]:
air_quality["london_mg_per_cubic"] = air_quality["station_london"] * 1.882
air_quality.head()

To create a new column, use the `[]` brackets with the new column name at the left side of the assignment.  The computation is done **element wise**, hence there is no need for an explicit loop!

We can also use multiple columns to derive a new column.

Let's check the ratio of the values in Paris versus Antwerp and save the result in a new column.

In [ ]:
air_quality["ratio_paris_antwerp"] = air_quality["station_paris"] / air_quality["station_antwerp"]
air_quality.head()

We can also perform more general mappings on columns, either using a function or a dictionary.

Let's work with the Titanic data once more.  Let's say we want to numerically encode the `Sex` column, this can be done as follows. First let's check the unique values in the `Sex` column.

In [ ]:
titanic['Sex'].unique()

In [ ]:
mapping_dict = {'male' : 0, 'female' : 1}
titanic['Sex'] = titanic['Sex'].map(mapping_dict)

Let's check whether it worked.

In [ ]:
titanic.head()

Let's add a new feature (i.e. a new column). When `Age` is less than 12 we will call this passenger a `child`, 
between 12 and 18 a `teen` and over 18 will be `adult`s.  First, we define a Python function implementing 
this mapping, and then we apply it to the `Age` column.

In [ ]:
def age_to_category(age):
    if age < 12:
        return "child"
    if age < 18:
        return "teen"
    return "adult"

titanic['AgeCategory'] = titanic['Age'].map(age_to_category)
titanic.head()

Here, we only scratched the surface of what is possible with pandas (and indeed, whole books have been written about pandas). In general the pandas getting started guide, the user guide and the API reference are very good resources to find additional information. You can find 
all of these [here](https://pandas.pydata.org/pandas-docs/stable/index.html).